File: src/web_server/html_render.rs
use maud::{html, Markup};

use crate::web_server::state::{Distribution, PriceStats};

use super::conditions::{CheapCondition, Condition, Eval, EvaluateContext};

pub fn render_layout(content: Markup) -> Markup {
    html! {
        html {
            head {
                title { "OTE CR Price Checker" }
                script src="https://unpkg.com/@tailwindcss/browser@4" {}
                script src="https://unpkg.com/htmx.org@2.0.4" integrity="sha384-HGfztofotfshcF7+8n44JQL2oJmowVChPTg48S+jvZoztPfvwD79OC/LTtG6dMp+" crossorigin="anonymous" {}
                script defer src="https://cdn.jsdelivr.net/npm/alpinejs@3.x.x/dist/cdn.min.js" {}
            }
            body .p-4.text-center."dark:bg-gray-900"."dark:text-gray-300" {
                (content)
            }
        }
    }
}

pub struct ChartSettings {
    pub height: f32,
    pub bar_width: usize,
    pub bar_spacing: usize,
}

impl Default for ChartSettings {
    fn default() -> Self {
        Self {
            height: 300.0,
            bar_width: 24,
            bar_spacing: 1,
        }
    }
}

struct ChartMetrics {
    scale: f32,
    zero_offset: f32,
    svg_width: usize,
    svg_height: f32,
}

impl ChartSettings {
    fn calculate_metrics(&self, prices: &[f32]) -> ChartMetrics {
        let cheapiest_hour = PriceStats::cheapest_hour(&prices);
        let expensive_hour = PriceStats::expensive_hour(&prices);

        let scale = if *cheapiest_hour.1 < 0.0 {
            self.height / (expensive_hour.1 - cheapiest_hour.1)
        } else {
            self.height / expensive_hour.1
        };

        let zero_offset = if *cheapiest_hour.1 < 0.0 {
            15.0 + (expensive_hour.1 * scale)
        } else {
            self.height + 15.0
        };

        ChartMetrics {
            scale,
            zero_offset,
            svg_width: prices.len() * (self.bar_width + self.bar_spacing),
            svg_height: self.height + 30.0,
        }
    }

    fn calculate_bar_x(&self, hour: usize) -> usize {
        hour * (self.bar_width + self.bar_spacing)
    }

    fn calculate_bar_y(&self, price: f32, metrics: &ChartMetrics) -> f32 {
        if price >= 0.0 {
            metrics.zero_offset - (price * metrics.scale)
        } else {
            metrics.zero_offset
        }
    }

    fn calculate_bar_height(&self, price: f32, metrics: &ChartMetrics) -> f32 {
        1.0_f32.max(price.abs() * metrics.scale)
    }

    fn calculate_text_x(&self, hour: usize) -> usize {
        hour * (self.bar_width + self.bar_spacing) + self.bar_width / 2
    }

    fn calculate_price_text_y(&self, price: f32, metrics: &ChartMetrics) -> f32 {
        metrics.zero_offset - (price * metrics.scale) - 3.0
    }

    fn calculate_label_text_y(&self, metrics: &ChartMetrics) -> f32 {
        metrics.zero_offset - 10.0
    }

    pub fn render(
        &self,
        prices: &[f32],
        labels: Option<&[&str]>,
        color: impl for<'a> Fn(&'a (usize, f32)) -> &'a str,
    ) -> Markup {
        let metrics = self.calculate_metrics(prices);

        html! {
            svg width=(metrics.svg_width) height=(metrics.svg_height) {
                g {
                    @for (hour, &price) in prices.iter().enumerate() {
                        rect x=(self.calculate_bar_x(hour)) y=(self.calculate_bar_y(price, &metrics))
                            width=(self.bar_width) height=(self.calculate_bar_height(price, &metrics))
                            class=(color(&(hour, price))) {}
                        text x=(self.calculate_text_x(hour)) y=(self.calculate_price_text_y(price, &metrics)) text-anchor="middle" .font-mono.text-xs."dark:fill-gray-300" {
                            (format!("{price:.0}"))
                        }

                        @if let Some(labels) = labels {
                            text x=(self.calculate_text_x(hour)) y=(self.calculate_label_text_y(&metrics)) text-anchor="middle" .font-mono.text-xs."dark:fill-gray-100" {
                                (labels[hour])
                            }
                        }
                    }
                }
            }
        }
    }
}

impl Condition {
    pub fn evaluate_all_in_chart(&self, ctx: &EvaluateContext) -> Markup {
        let results = self.evaluate_all(ctx);

        let labels = &results
            .iter()
            .map(|result| if *result { "T" } else { "F" })
            .collect::<Vec<&str>>();

        let chart = ChartSettings::default();
        chart.render(&ctx.prices.prices, Some(&labels), |(index, _price)| {
            if results[*index] {
                "fill-green-600"
            } else {
                "fill-red-600"
            }
        })
    }
}

fn format_price(price: f32) -> Markup {
    html! {
        (price.floor())
        span .text-neutral-500 .text-sm {
            "."(format!("{:02.0}", (price - price.floor()) * 100.0 ))
        }
    }
}

pub fn link(url: &str, text: &str) -> Markup {
    html! {
        a href=(url) a .underline ."hover:text-red-400" { (text) }
    }
}

impl crate::web_server::state::DayPrices {
    pub(crate) fn render_table(&self, dist: &Distribution) -> Markup {
        let total_prices = self.total_prices(dist);

        let (_, &total_low) = PriceStats::cheapest_hour(&&(total_prices[..]));
        let (_, &total_high) = PriceStats::expensive_hour(&&(total_prices[..]));

        html! {
            table {
                tr {
                    th.pr-10 { "Hour" }
                    th colspan="2" { "Price EUR/MWh" }
                }
                tr {
                    th.pr-10 { "" }
                    th.pr-10 { "Market" }
                    th { "With Distribution" }
                }
                @for (hour, &price) in self.prices.iter().enumerate() {
                    tr
                        ."bg-green-100"[total_prices[hour] == total_low]
                        ."dark:bg-green-900"[total_prices[hour] == total_low]
                        .bg-red-100[total_prices[hour] == total_high]
                        ."dark:bg-red-900"[total_prices[hour] == total_high]
                    {

                        td .text-right .font-mono .pr-10 {
                            (hour)
                            span .text-neutral-500 .text-sm {
                                " : 00 - 59"
                            }
                        }
                        td .text-right .text-green-700[price<0.0] .font-mono .pr-10 {
                            (format_price(price))
                        }
                        td .text-right .text-green-700[price<0.0] .font-mono {
                            (format_price(total_prices[hour]))
                        }
                    }
                }
            }
        }
    }
}

pub trait RenderHtml {
    fn render_html(&self) -> Markup;
}

impl RenderHtml for Condition {
    fn render_html(&self) -> Markup {
        match self {
            Condition::And(conditions) => html! {
                div .ml-4 {
                    "AND"
                    ul {
                        @for condition in conditions {
                            li { (condition.render_html()) }
                        }
                    }
                }
            },
            Condition::Or(conditions) => html! {
                div .ml-4 {
                    "OR"
                    ul {
                        @for condition in conditions {
                            li { (condition.render_html()) }
                        }
                    }
                }
            },
            Condition::Not(condition) => html! {
                div .ml-4 {
                    "NOT"
                    (condition.render_html())
                }
            },
            Condition::Price(price) => html! {
                div .ml-4 {
                    "Price: " (price)
                }
            },
            Condition::Hours(from, to) => html! {
                div .ml-4 {
                    "Hours: " (from) " - " (to)
                }
            },
            Condition::Cheap(CheapCondition { hours, from, to }) => html! {
                div .ml-4 {
                    "Cheap: " (hours) " cheapiest hours in hours " (from) " - " (to)
                }
            },

            #[cfg(test)]
            Condition::Debug(_) => todo!(),
        }
    }
}

impl RenderHtml for Option<CheapCondition> {
    fn render_html(&self) -> Markup {
        let actual = self.as_ref().unwrap_or_else(|| &CheapCondition {
            hours: 1,
            from: 0,
            to: 24,
        });

        html! {
            form method="GET" class="flex space-x-2 items-center" {
                label for="cheap_hours" { "Cheap Hours:" }
                input type="number" id="cheap_hours" name="hours" value=(actual.hours) min="1" max="24" step="1" class="w-16 p-1 border rounded" {}
                label for="cheap_from" { "From:" }
                input type="number" id="cheap_from" name="from" value=(actual.from) min="0" max="23" step="1" class="w-16 p-1 border rounded" {}
                label for="cheap_to" { "To:" }
                input type="number" id="cheap_to" name="to" value=(actual.to) min="1" max="24" step="1" class="w-16 p-1 border rounded" {}
                button type="submit" class="px-4 py-1 bg-blue-500 text-white rounded cursor-pointer" { "Update" }
            }
        }
    }
}

#[cfg(test)]
mod chart_settings_tests {
    use super::*;

    #[test]
    fn test_chart_settings_with_prices_negative_zero_positive() {
        let settings = ChartSettings::default();
        let prices = vec![-10.0, 0.0, 10.0];
        let metrics = settings.calculate_metrics(&prices);

        // Verify scale calculation
        assert_eq!(metrics.scale, 15.0); // 300.0 / 20.0

        // Verify zero offset
        assert_eq!(metrics.zero_offset, 165.0); // 15.0 + (10.0 * 15.0)

        // Verify bar heights
        assert_eq!(settings.calculate_bar_height(-10.0, &metrics), 150.0); // same height as +10.0
        assert_eq!(settings.calculate_bar_height(0.0, &metrics), 1.0); // min height for zero
        assert_eq!(settings.calculate_bar_height(10.0, &metrics), 150.0); // 10.0 * 15.0

        // Verify bar Y positions
        assert_eq!(settings.calculate_bar_y(-10.0, &metrics), 165.0); // at zero line for negative
        assert_eq!(settings.calculate_bar_y(0.0, &metrics), 165.0); // at zero line for zero
        assert_eq!(settings.calculate_bar_y(10.0, &metrics), 15.0); // above zero line for positive
    }
}

File: webserver/htmlrender.go
package webserver

import (
	"embed"
	"fmt"
	"html"
	"html/template"
	"math"
	"strings"
)

const bodyClasses = "p-4 text-center dark:bg-gray-900 dark:text-gray-300"

//go:embed templates/*.html
var templatesFS embed.FS

var layoutTmpl = template.Must(template.ParseFS(templatesFS, "templates/*.html"))

// RenderLayout wraps a page body in the full HTML document and emits the generated CSS.
func RenderLayout(content string) string {
	classes := map[string]struct{}{}
	ExtractClassesFromHTML(content, classes)
	ExtractClassesFromStr(bodyClasses, classes)
	css := GenerateCSS(classes)

	var sb strings.Builder
	err := layoutTmpl.ExecuteTemplate(&sb, "layout.html", struct {
		CSS         template.CSS
		BodyClasses string
		Content     template.HTML
	}{
		CSS:         template.CSS(css),
		BodyClasses: bodyClasses,
		Content:     template.HTML(content),
	})
	if err != nil {
		panic(err)
	}
	return sb.String()
}

// ChartSettings controls bar chart dimensions.
type ChartSettings struct {
	Height     float32
	BarWidth   int
	BarSpacing int
}

func DefaultChartSettings() ChartSettings {
	return ChartSettings{Height: 300.0, BarWidth: 24, BarSpacing: 1}
}

type chartMetrics struct {
	scale      float32
	zeroOffset float32
	svgWidth   int
	svgHeight  float32
}

func (cs ChartSettings) calculateMetrics(prices []float32) chartMetrics {
	_, cheapest := CheapestHour(prices)
	_, expensive := ExpensiveHour(prices)

	var scale, zeroOffset float32
	if cheapest < 0 {
		scale = cs.Height / (expensive - cheapest)
		zeroOffset = 15.0 + expensive*scale
	} else {
		scale = cs.Height / expensive
		zeroOffset = cs.Height + 15.0
	}

	return chartMetrics{
		scale:      scale,
		zeroOffset: zeroOffset,
		svgWidth:   len(prices) * (cs.BarWidth + cs.BarSpacing),
		svgHeight:  cs.Height + 30.0,
	}
}

func (cs ChartSettings) calculateBarX(hour int) int {
	return hour * (cs.BarWidth + cs.BarSpacing)
}

func (cs ChartSettings) calculateBarY(price float32, m chartMetrics) float32 {
	if price >= 0 {
		return m.zeroOffset - price*m.scale
	}
	return m.zeroOffset
}

func (cs ChartSettings) calculateBarHeight(price float32, m chartMetrics) float32 {
	v := float32(math.Abs(float64(price))) * m.scale
	if v < 1.0 {
		return 1.0
	}
	return v
}

func (cs ChartSettings) calculateTextX(hour int) int {
	return hour*(cs.BarWidth+cs.BarSpacing) + cs.BarWidth/2
}

func (cs ChartSettings) calculatePriceTextY(price float32, m chartMetrics) float32 {
	return m.zeroOffset - price*m.scale - 3.0
}

func (cs ChartSettings) calculateLabelTextY(m chartMetrics) float32 {
	return m.zeroOffset - 10.0
}

// Render returns the SVG markup for a bar chart over prices.
func (cs ChartSettings) Render(prices []float32, labels []string, color func(index int, price float32) string, currency Currency) string {
	metrics := cs.calculateMetrics(prices)

	var sb strings.Builder
	fmt.Fprintf(&sb, `<svg viewBox="0 0 %d %s" style="max-width:%dpx">`,
		metrics.svgWidth, fmtFloat(metrics.svgHeight), metrics.svgWidth)
	sb.WriteString("<g>")
	for hour, price := range prices {
		cls := color(hour, price)
		fmt.Fprintf(&sb, `<rect x="%d" y="%s" width="%d" height="%s" class="%s" data-idx="%d"></rect>`,
			cs.calculateBarX(hour),
			fmtFloat(cs.calculateBarY(price, metrics)),
			cs.BarWidth,
			fmtFloat(cs.calculateBarHeight(price, metrics)),
			cls,
			hour,
		)
		var priceStr string
		if currency == CurrencyCzk {
			priceStr = fmt.Sprintf("%.1f", currency.Convert(price))
		} else {
			priceStr = fmt.Sprintf("%.0f", currency.Convert(price))
		}
		fmt.Fprintf(&sb, `<text x="%d" y="%s" text-anchor="middle" class="font-mono text-xs dark:fill-gray-300">%s</text>`,
			cs.calculateTextX(hour),
			fmtFloat(cs.calculatePriceTextY(price, metrics)),
			html.EscapeString(priceStr),
		)
		if labels != nil {
			fmt.Fprintf(&sb, `<text x="%d" y="%s" text-anchor="middle" class="font-mono text-xs dark:fill-gray-100">%s</text>`,
				cs.calculateTextX(hour),
				fmtFloat(cs.calculateLabelTextY(metrics)),
				html.EscapeString(labels[hour/4]),
			)
		}
	}
	sb.WriteString("</g></svg>")
	return sb.String()
}

func fmtFloat(f float32) string {
	// Print without trailing zeros, mimicking Rust's f32 Display.
	s := fmt.Sprintf("%g", f)
	return s
}

// EvaluateAllInChart renders a chart visualizing condition results across the context.
func (c Condition) EvaluateAllInChart(ctx *EvaluateContext) string {
	results := c.EvaluateAll(ctx)
	labels := make([]string, len(results))
	for i, r := range results {
		if r {
			labels[i] = "T"
		} else {
			labels[i] = "F"
		}
	}
	chart := DefaultChartSettings()
	return chart.Render(ctx.Prices.Prices, labels, func(index int, _ float32) string {
		if results[index] {
			return "fill-green-600"
		}
		return "fill-red-600"
	}, CurrencyEur)
}

func formatPrice(price float32, currency Currency) string {
	s := fmt.Sprintf("%.2f", currency.Convert(price))
	idx := strings.Index(s, ".")
	return fmt.Sprintf(`%s<span class="text-neutral-500 text-sm">.%s</span>`,
		s[:idx], s[idx+1:])
}

// Link returns an anchor tag with the underline+hover style.
func Link(url, text string) string {
	return fmt.Sprintf(`<a href="%s" class="underline hover:text-red-400">%s</a>`,
		html.EscapeString(url), html.EscapeString(text))
}

// RenderTable renders the 24x4 table of quarter-hour prices for a single day.
func (d *DayPrices) RenderTable(dist *Distribution, actualIndex int, currency Currency, includeDist bool) string {
	totalPrices := d.TotalPrices(dist)
	var displayPrices []float32
	if includeDist {
		displayPrices = totalPrices
	} else {
		displayPrices = d.Prices
	}

	minIdx, _ := CheapestHour(displayPrices)
	maxIdx, _ := ExpensiveHour(displayPrices)

	var sb strings.Builder
	sb.WriteString("<table>")
	sb.WriteString(`<tr><th class="px-4">Hour</th><th class="px-4">:00</th><th class="px-4">:15</th><th class="px-4">:30</th><th class="px-4">:45</th></tr>`)
	for hour := 0; hour < 24; hour++ {
		sb.WriteString("<tr>")
		fmt.Fprintf(&sb, `<td class="text-right font-mono px-4">%d:00</td>`, hour)
		for q := 0; q < 4; q++ {
			idx := hour*4 + q
			price := displayPrices[idx]
			classes := []string{"text-right", "font-mono", "px-4"}
			if idx == minIdx {
				classes = append(classes, "bg-green-100", "dark:bg-green-900")
			}
			if idx == maxIdx {
				classes = append(classes, "bg-red-100", "dark:bg-red-900")
			}
			if idx == actualIndex {
				classes = append(classes, "font-bold", "outline-2", "outline-blue-500")
			}
			if price < 0 {
				classes = append(classes, "text-green-700")
			}
			fmt.Fprintf(&sb, `<td class="%s" data-idx="%d">%s</td>`,
				strings.Join(classes, " "), idx, formatPrice(price, currency))
		}
		sb.WriteString("</tr>")
	}
	sb.WriteString("</table>")
	return sb.String()
}

// RenderHTML returns the HTML representation of the condition tree.
func (c Condition) RenderHTML() string {
	switch c.Kind {
	case CondAnd:
		var sb strings.Builder
		sb.WriteString(`<div class="ml-4">AND<ul>`)
		for _, child := range c.Children {
			sb.WriteString("<li>")
			sb.WriteString(child.RenderHTML())
			sb.WriteString("</li>")
		}
		sb.WriteString("</ul></div>")
		return sb.String()
	case CondOr:
		var sb strings.Builder
		sb.WriteString(`<div class="ml-4">OR<ul>`)
		for _, child := range c.Children {
			sb.WriteString("<li>")
			sb.WriteString(child.RenderHTML())
			sb.WriteString("</li>")
		}
		sb.WriteString("</ul></div>")
		return sb.String()
	case CondNot:
		return `<div class="ml-4">NOT` + c.Inner.RenderHTML() + `</div>`
	case CondPrice:
		return fmt.Sprintf(`<div class="ml-4">Price: %s</div>`, fmt.Sprintf("%g", c.Price))
	case CondHours:
		return fmt.Sprintf(`<div class="ml-4">Hours: %d - %d</div>`, c.HoursMin, c.HoursMax)
	case CondCheap:
		return fmt.Sprintf(`<div class="ml-4">Cheap: %d cheapiest hours in hours %d - %d</div>`,
			c.Cheap.Hours, c.Cheap.From, c.Cheap.To)
	}
	return ""
}

// RenderCheapForm renders the GET form for editing the Cheap fields.
func RenderCheapForm(cc *CheapCondition) string {
	actual := cc
	if actual == nil {
		actual = &CheapCondition{Hours: 1, From: 0, To: 24}
	}
	var sb strings.Builder
	sb.WriteString(`<form method="GET" class="flex space-x-2 items-center">`)
	sb.WriteString(`<label for="cheap_hours">Cheap Hours:</label>`)
	fmt.Fprintf(&sb, `<input type="number" id="cheap_hours" name="hours" value="%d" min="1" max="24" step="1" class="w-16 p-1 border rounded">`, actual.Hours)
	sb.WriteString(`<label for="cheap_from">From:</label>`)
	fmt.Fprintf(&sb, `<input type="number" id="cheap_from" name="from" value="%d" min="0" max="23" step="1" class="w-16 p-1 border rounded">`, actual.From)
	sb.WriteString(`<label for="cheap_to">To:</label>`)
	fmt.Fprintf(&sb, `<input type="number" id="cheap_to" name="to" value="%d" min="1" max="24" step="1" class="w-16 p-1 border rounded">`, actual.To)
	sb.WriteString(`<button type="submit" class="px-4 py-1 bg-blue-500 text-white rounded cursor-pointer">Update</button>`)
	sb.WriteString(`</form>`)
	return sb.String()
}
